traitlets/traitlets.py

```diff
@@ -390,20 +385,18 @@ def _setup_dynamic_initializer(self, obj):
         # trait declaration or above.
         mro = type(obj).mro()
         meth_name = '_%s_default' % self.name
-        for cls in mro[:mro.index(self.this_class)+1]:
+        for cls in mro[:mro.index(self.this_class) + 1]:
             if meth_name in cls.__dict__:
                 break
         else:
-            return False
-        # Complete the dynamic initialization.
+            return
         obj._trait_dyn_inits[self.name] = meth_name
-        return True
 
     def _set_default_value_at_instance_init(self, obj):
         # As above, but if no default was specified, don't try to set it.
         # If the trait is accessed before it is given a value, init_default_value
         # will be called at that point.
-        if (not self._setup_dynamic_initializer(obj)) \
+        if (self.name not in obj._trait_dyn_inits) \
```

Comment:
- #28 does not change behavior. It is more of a cosmetic change and could be merge more easily.
- #23 is more involving since it changes how default values are initialized.